src/AspNet.Security.OAuth.Introspection/Events/BaseIntrospectionContext.cs

```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+
+namespace AspNet.Security.OAuth.Introspection
+{
+    public abstract class BaseIntrospectionContext : BaseControlContext
```

Comment:
Initially, `BaseControlContext` had been designed to provide a convenient way to determine whether a middleware had to continue the request processing, should terminate the request immediately, or should directly jump to the next middleware, nothing more.

This base class has been (abusively IMHO) extended to pretty much every event, even those that are not responsible of the request processing and shouldn't alter the response, which is true for all the events called from `HandleAuthenticateAsync`, because this method can be called from user code (`HttpContext.AuthenticationManager.AuthenticateAsync`). Here (just like in the JWT bearer middleware), the events don't really control the HTTP request/response since we can't prevent the next middleware in the pipeline from being invoked (`HandleResponse` just skips the defaut logic, but the next middleware is still called).

ASOS used to suffer from the same symptom, but I reworked all the events to use `BaseControlContext` only where we can really return an HTTP response (basically, all the `HandleXXXRequest`/`ApplyXXXResponse` events, but not `ValidateXXXRequest` or `Serialize/DeserializeXXX`).